### ISSUE TITLE
Breaking change: `x-changed` event fired before internal bindings dependent on `x` are run.

### ISSUE THREAD

**Issue by kevinpschaaf** [Polymer/polymer#3910]
_From @bicknellr on July 27, 2016 17:47_

It's no longer safe to assume that after receiving a change notification event (`x-changed`) from an element that all of that element's internal observers that are dependent on the changed value (`x`) have run.

For example, one of the tests for iron-image waits on `loaded-changed` to check that the iron-image's placeholder is hidden. Because the placeholder's visibility is determined by a multi-property binding to its class attribute, the `loaded-changed` event is now fired _before_ the placeholder's class list has been updated causing the test (which uses `getComputedStyle`) to inspect styles too early and fail.

_Copied from original issue: PolymerLabs/alacarte#12_

**Comment by kevinpschaaf** [comment]
Just a correction, in 1.0 _`observer`s_ did not run before `notify`, but bindings (`annotation` effects) did.  So depending on how the effect of a property change was implemented, it may or may not have been "flushed" by the time the event was fired.

**Comment by kevinpschaaf** [comment]
Will revisit during "more BC" work on property-effects layer.

**Comment by bicknellr** [comment]
Any updates?

**Comment by kevinpschaaf** [comment]
Leaning toward leaving existing 2.0 behavior; in 1.x, the order of effects was

1. `computed`
1. `bindings`
1. **`notify`**
1. `observers`

In 2.0-preview, it is now:

1. `computed`
1. **`notify`**
1. `bindings`
1. `observers`

This ensures property changes are notified to the top-most element first, then propagated once back down the tree, resulting in more deterministic ordering and avoiding the need for dirty-checking to avoid duplicate work, which is required to remove object/array dirty-checking.

It is a breaking change in that the notification now happens before bindings, but it was never particularly safe to assume that all changes as a result of the property being notified have completed since observers always ran after the notification (which could have also resulted in changes to the host and/or more binding to occur).  Since the breaking change is beneficial in many ways, and only weakens an already weak guarantee, we're leaning towards keeping the change and clearly documenting that "automatic `notify:true` notifications occur before changes have been processed, meaning it is a notification that a state change will occur, not that its side effects have been completed".

This means we would need to consider alternate patterns for the e.g. measurement/testing case.  The most straightforward is to consider `notify:true`-based `property-changed` events as being used for data-binding usage between hosts/children (with "pre-change" semantics), and if an element wants to fire an event indicating some type of side-effects have occurred, it can fire a custom event.

**Comment by sorvell** [comment]
Same basic issue but different manifestation here: https://github.com/Polymer/polymer/issues/3932.

**Comment by notwaldorf** [comment]
I'm just going to leave this JSBin here: http://output.jsbin.com/hubacez
TL; DR: when `input` fires its events, it's already processed them internally, so this 2.0 behaviour isn't even what the platform does :(

**Comment by sorvell** [comment]
Fixed via https://github.com/Polymer/polymer/pull/4221.

The effect order was reworked as follows:
1. `computed`
2. `bindings`
3. `observers`
4. **`notify`**
